### DISCUSSION TITLE
Table attributes: row height, row merge, row background color

### DISCUSSION THREAD

**Discussion by hoffenmerCode** [beeware/toga#1292] [Q&A]
Are these possible?

1. change row height or column width
2. merge rows or columns
3. change background color of table cells

If yes, how? style=Pack doesn't work for the above. Please help.

**Comment by freakboy3742** [comment]
Generally speaking, no. Tables are native widgets, so they adopt the operating system-mandated style. 

This means features like row height isn't under the control of the end-user; they're a function of how the OS draws content in the table. 

There has been some discussion about specifying initial column widths; however, those API discussions haven't been completed. Cell background color could potentially fit into those discussions, but hasn't been to date.

As for merging rows/columns - while I'm familiar with the practice in the context of HTML tables, I'm not sure how that would map into a native widget. Although they share a name, they're not really the same concept. 

More generally, you should be wary of "how can I do this HTML thing in Toga" lines of reasoning. Although we're using a CSS-like language for configuration, Toga isn't HTML - by design. We've deliberately restricted what can be done visually to preserve *native* look and feel, and we'll be hesitant to introduce features that allow developers to work around that design choice.

**Reply by hoffenmerCode** [reply]
Thanks for the very prompt reply. 
What about setting default width of a splitContainer? Instead of 50/50 say 20/80?

**Reply by freakboy3742** [reply]
The answer for SplitContainer is a little different; it is possible to specify the size/weight of the split when you specify the content for the container:

    my_split = toga.SplitContainer(content=[
        (right_content, 0.2),
        (left_content, 0.8),
    ])

Arguably this *should* be specified with a style parameter, but for the moment at least, this approach works.

**Reply by hoffenmerCode** [reply]
Again, really appreciate your input and prompt responses. Few more questions: 

1. re SplitContainer = how can I search for how these widgets can be changed? It is not documented online. I tried reading the source code/docs but can't find the details you're providing. Any advice? So I don't bother you too much.

2. Is there a toggle widget in the works/future? I currently use/group "Switch" and enable/disable them. A toggle would be better solution I think. 

3. OSX desktop applications = I can't get the time picker or date picker to work. Can't even run the program, I get an error that said module is not known. 

4. What is the future of beeware/toga? I gave up on QT as I prefer toga hands down but would like to know whether the plan is to make a full featured stable version for production-level programming? I use toga for OSX desktop applications.

**Reply by hoffenmerCode** [reply]
Also, the SplitContainer sizing doesn't work for me on OSX: 

`mainSplit = toga.SplitContainer()
        splitLeftSearchContainer = toga.SplitContainer(direction=False)
        splitRightCompanyContainer = toga.SplitContainer(direction=False)

        splitLeftSearchContainer.content=[(containerC, 0.4), (containerD,0.6)]
        splitRightCompanyContainer.content=[containerE, containerF]
        mainSplit.content = [(splitLeftSearchContainer, 0.4),
                                (splitRightCompanyContainer, 0.6)]`

Ideas?

**Reply by freakboy3742** [reply]
1. Toga is v0.3. Documentation is incomplete, and there are bugs. There's no magic solution for this, and no magic supply of hidden documentation.
2. One of the design ideas behind Toga is that GUIs should follow the UX guidelines for the platform. As a result, the API that Toga exposed doesn't map directly to specific widgets - it tries to capture "concepts". The Switch widget, for example, manifests as a checkbox on macOS, because that's what the macOS HIG recommends for boolean interactions. On iOS, it's a switch. So no - there's no plan to introduce a "toggle" widget, because some platforms (including macOS) advice *against* the use of that widget for boolean choices.
3. Yes. You're seeing an error that says the widget doesn't exist. That's because it hasn't been written yet.
4. The end goal is for Toga to be a product-capable GUI toolkit. However, that's obviously a big task, and we're only at v0.3 at present. That means there are bugs. 

As for the SplitContainer widget - that sounds like a bug; it *should* be implemented on macOS. When you say "doesn't work" do you mean you're seeing an error, or that the behavior isn't changing?

**Reply by hoffenmerCode** [reply]
There is no error. The behavior is not changing.

**Reply by freakboy3742** [reply]
Interesting... looks like something isn't working with the macOS implementation. It definitely works on GTK. I'll need to investigate further.

**Reply by freakboy3742** [reply]
I've got a fix for this in #1300. Thanks for the report!

**Comment by royccg** [comment]
Is it possible to change the color of individual rows in the table now?

**Reply by freakboy3742** [reply]
No - there has been no change to add color features to individual rows.
